### DISCUSSION TITLE
Can I use two backends in two filters?

### DISCUSSION THREAD

**Discussion by SSShooter** [fabricjs/fabric.js#9778] [Q&A]
I want to do a histogram equalization in a filter which processes the `imageData`, on the other hand I have a filter use a fragment shader.

But it seem use webgl backend by default. Can I use two backends at the same time?

**Comment by asturur** [comment]
No you can't use 2 backends, at time, you have to choose one.
What does it stops you to write both filter in the same way?

**Reply by SSShooter** [reply]
I want to process `imageData` using convolution and perform a histogram equalization.
As far as I know, I can't access `imageData` through webGL backend.

**Reply by asturur** [reply]
well you can, otherwise you woun't be able to do any filter, correct?
I m just not sure how hard is to change the code from imageData to webgl, if the code is not secret and you want to share it here we can talk about that. I m not great at webgl but i wrote the other filters

**Reply by SSShooter** [reply]
Thank you! 
Please see https://github.com/SSShooter/canvas-img-process/blob/e41d752e9bbecb075af47a3d226e765e075f19a0/src/App.vue#L111
I want to access pixels which surround the current pixel.

**Reply by asturur** [reply]
it seems to me this is just a convolution or colorMatrix filters as we have implemented in fabricJS.

https://github.com/fabricjs/fabric.js/blob/master/src/filters/ColorMatrix.ts
https://github.com/fabricjs/fabric.js/blob/master/src/filters/Convolute.ts

With one of those 2 you can implement that exact kernel
Please take your time to understand the difference between a color matrix and a convolution, once you have a filter setup passing a different image is quite easy you can try easily different combinations

**Reply by SSShooter** [reply]
Thanks for your information! It's really helpful! 🤩
I thought I can't get pixel around a specific pixel, till I found [this](https://github.com/fabricjs/fabric.js/blob/2953be4cb35dca41bedce9115673be1ced561d32/src/filters/shaders/convolute.ts#L90).

**Reply by asturur** [reply]
not sure why the histogram requires nearby pixels, but let me know how it goes.
The convolute filter can be used right away with different matrices you can pass in, and if you need more matrices one after the other you can apply many times the convolute filter

**Reply by SSShooter** [reply]
oh, I mean convolution requires nearby pixels. histogram eq, it seems, doesn't need them.
I struggle finding a way to convert  [this](https://github.com/tpmccauley/histogram-equalization/blob/main/js/image-hist.js) to a shader version.

**Reply by asturur** [reply]
maybe here you can take some info
https://jsfiddle.net/greggman/9amgpndt/
